### ISSUE TITLE
Allow JX-serialization of cyclic objects

### ISSUE THREAD

**Issue by fatcerberus** [svaarala/duktape#331]
I know JSON specifically doesn't allow cyclic objects to be serialized, but it would be useful if JX could do so.  Obviously round-trip serialization would be impossible, but maybe Duktape could put some sort of backreference in the output when it encountered a cycle, similiar to how it adds `{ _func:true }` tags for otherwise unserializable functions.

Currently variable viewing in the minisphere debugger is limited because there are no commands to walk objects, so I have to rely on `Duktape.enc` and an Eval command to display the full object.  This fails, however, in the case of a cyclic object (which it turns out happens _a lot_ in the game I'm making).

**Comment by svaarala** [comment]
This should be relatively simple to implement. Here's a quick hack branch of one approach (it's a hack in that it breaks JSON.stringify() output for example):

- https://github.com/svaarala/duktape/compare/cyclic-jx-gh331

Example:

```
$ ./duk
((o) Duktape 1.2.99 (v1.2.0-583-g81c2e5d-dirty)
duk> obj = {};
= [object Object]
duk> obj.foo = {}
= [object Object]
duk> obj.foo.bar = {}
= [object Object]
duk> obj.foo.bar.quux = { ref: obj.foo }
= [object Object]
duk> Duktape.enc('jx', obj)
= {foo:{bar:{quux:{ref:{_loop:true}}}}}
```

The plumbing is a bit awkward so maybe reorganizing the helpers a bit would be needed.

**Comment by fatcerberus** [comment]
Even more useful would be if there were some way for the _loop tag to specify what it looped back on, but that would probably be too difficult to implement and/or hurt performance.

As long as all the unique objects are present in the output I'll be happy. :)

**Comment by fatcerberus** [comment]
Looked at the diff, I'm impressed at how few changes were needed for a minimal implementation.  I thought this would be a lot more difficult!

**Comment by svaarala** [comment]
One possible format would be something like:

```
{_loop:(0xdeadbeef)} 
```

The pointer there could be the `duk_heaphdr` address of the object. But since it's not present in the serialized form of the target object, this only really serves to indicate if there are multiple separate loops.

A more fancy version would be to serialize the `_loop` tag and then perhaps serialize the target shallowly (just one level) which would mostly allow identifying the target. For example if the target object has an `.id` or `.type` property or similar, that would be shown. But this will essentially serialize certain objects twice so it seems quite messy to me.

**Comment by fatcerberus** [comment]
I was thinking more a "path" of sorts to identify the loop (e.g. `_loop:"foo.bar"`).  This would always be meaningful since by definition a cycle is self-contained.

**Comment by fatcerberus** [comment]
Oh, and additionally, it would allow deserialization with the circular refs intact.

**Comment by svaarala** [comment]
Oh, I see. That information would need a walk through the loop to get the property name list.

This is not as easy as it seems, because there may be multiple "branches" in the walk path and multiple independent loops.
```

**Comment by svaarala** [comment]
It would also be quite interesting with accessors where every time you tried to walk the property chain you might get a different result from each accessor. Corner case, sure, but the JSON code needs to deal with all cases.

**Comment by fatcerberus** [comment]
Yeah, like I said it's probably not possible to implement without some major changes under the hood.  Just figured I'd throw it out there. :-)

**Comment by svaarala** [comment]
By the way, regarding the debugger heap walking, I think the easiest approach to ensure safety of heap walking would be to prevent both mark-and-sweep passes and refzero queue freeing while the debugger is in paused state. That way all heap object references would always point to valid memory (even when their refcounts dropped to zero during debugging - e.g. for the result of an `Eval` command).

The downside would of course be that you could run out of memory while paused if you did too many Evals for example.

**Comment by svaarala** [comment]
Actually, getting the property name list would probably be possible through the loop detection stack; it already stores the objects being processed (for the slow path this is a hash map but it could also have an additional array). It would then need to store the keys through which the loop stack has been accumulated. This would then give the property list when one walked back the stack until the loop target was found in the stack.

But yeah, it'd be quite a lot of work and unnecessary book-keeping when one didn't specifically want to have it.

**Comment by svaarala** [comment]
Pushed a slightly better hack which doesn't break JSON (should only affect JX):

```
$ ./duk
((o) Duktape 1.2.99 (v1.2.0-584-g75705c8-dirty)
duk> obj1 = {}; obj2 = {}; obj1.ref1 = obj2; obj2.ref1 = obj1; obj2.ref2 = obj1;
= [object Object]
duk> Duktape.enc('jx', obj1)
= {ref1:{ref1:{_loop:true},ref2:{_loop:true}}}
duk> JSON.stringify(obj1)
TypeError: cyclic input
	duk_bi_json.c:1432
	stringify  native strict preventsyield
	global input:1 preventsyield
```

**Comment by fatcerberus** [comment]
Agreed on suspending GC during debugger pause.  With enough ingenuity it would probably be possible to segfault the target through the debugger, which is obviously not optimal. ;)

**Comment by svaarala** [comment]
It doesn't need ingenuity really - if you just use Eval to arrive at a result object, that object won't be reachable once the Eval is complete. If Eval was changed to provide a "walkable reference" it would immediately have a stale pointer if refzero queue was freed.

A less invasive approach would be to have commands to "enable walking" and "disable walking" which would control the gc freeze. Slightly tedious for the debug client, but would limit the out-of-memory risks only when walking was actually done.

**Comment by fatcerberus** [comment]
Alternatively, there could be a command the client could send that would be equivalent to `Duktape.gc()`, which could be requested every once in a while to drive the GC and prevent running out of memory.

Anyway, this should probably go in a separate issue.

**Comment by fatcerberus** [comment]
Just tried this out, it seems to work well.  I was quite surprised to see where the loops in the object were, they weren't where I expected them to be!  This is very useful for debugging.

It did serve to illustrate how much JX encoding would benefit from the recent JSON speedups, though, the objects in question (representing battlers in an RPG) turned out to be quite large and took a few seconds to serialize.

**Comment by fatcerberus** [comment]
Wow, so either there's a bug in this causing it to not detect a loop properly, or the object I'm trying to view is hella complex--when attempting to view one the above-mentioned battler objects, minisphere's RAM usage shot up to nearly 2GB for about almost a full minute, and then the debug client locked up the machine trying to deal with what must have been an enormous string... :stuck_out_tongue_closed_eyes:

**Comment by svaarala** [comment]
There might be e.g. a value stack handling bug there, it was just a quick test.

But it's quite easy to generate much larger output than input with just loop detection in place. For example, suppose you have some object that serializes to 10kB of JX output. Then create a list of 1000 such objects (1000 separate references to that object). While the heap is not much larger (the references take around 16kB), the output is 1000x larger, i.e. 10MB. Then create another list that references that 1000-object list. Each reference there costs 16 bytes of RAM but generates 10MB of output. There are no loops - each object is just visited a lot of times.

This is actually why loop detection alone is not very useful IMO for debugging or logging values (I'll post a separate note on that). That's my experience anyway, with just loop detection you'll sometimes get ridiculously large output because of some accidental reference somewhere which causes some object to reference some self-contained but huge object (say some config or state object).

Anyway, for debugging, heap walking would be the best approach because the debugger could then allow the user to expand values, trigger getter calls, etc.

**Comment by svaarala** [comment]
Anyway, for JX formatting I had roughly the following option list:

- Detecting loops and stubbing them as `{_loop:true}`
- Detecting a depth limit and stubbing objects beyond that limitas `{_depth:true}`
- Detecting visited objects and avoiding their re-serialization (even if they're outside loops), stubbing them as `{_visited:true}`
- Avoiding getter calls and stubbing them as `{_getter:true}`
- Avoiding Proxy traps and stubbing them as `{_proxy:true}`
- Avoid `.toJSON()` calls, as they cause side effects
- Truncate large strings and buffers to specified limit, indicate truncation and real length, e.g. `|deadbeef|(256)`.
- Serialize strings as raw byte data (allows exact serialization of non-UTF-8 strings)
- Detect a maximum serialization output length, and truncate *whole output* at that point (basically throw an error, but keep the output, and stick a `...` after)
- Serialize non-enumerable properties too; but perhaps only of the level 1 object and not of parents (might need an option for parents or depth limit for non-enumerable properties)

In a project unrelated to Duktape we have most of these incorporated in our serializer, and it's quite useful as the serializer is usable in much more situations than plain JSON would be. But it does mean quite a lot of options, because these contexts are quite different.

There's a place for these in `Duktape.enc()` so that one could e.g.:

```js
Duktape.enc('jx', val, {
    loop: true,   // detect loops
    depth: 3,     // restrict to max depth 3
    size: 4096,   // limit max output to 4kB
});
```

This would then also need to be available for the C API (if one is added) where things get quite annoying: representing those options in a version extensible C API is a bit of an issue.

I originally looked into this for logging: it would be very useful if no side effects were triggered and there was a guaranteed maximum output size (it's rarely useful for a log entry to be beyond a few kB longer, even if legitimate). For JX to be truly useful for logging, one will need most of the above options to cover all relevant cases occurring in application code.

I left these out from Duktape 1.0 originally because there's some footprint impact. Also some features need new tracking state (like detecting "visited" objects and not just "looped" references). Finally, I didn't have time to figure out the C API for accessing these options (using a boxed option object seemed lazy :) and ended up leaving JX/JC out of the C API at that time.

**Comment by fatcerberus** [comment]
I'll have to look into the objects in question to see exactly what properties they maintain.  If I really am generating 2GB of data by serializing it, that's going to be an issue for game saving. :)

**Comment by svaarala** [comment]
The only reason a non-cyclic structure would be "amplified" if some objects get repeatedly serialized, i.e. they are referenced multiple times. That's obviously a problem for saving game state, because when you load it back each such re-serialization will be loaded back into a separate object which is probably not intended.

If the are multiple references to the same object for some legitimate reason, making all but one of the references non-enumerable causes JSON and JX to skip the property during serialization.

**Comment by svaarala** [comment]
A "serialize anything to a useful string" format (JX with some changes) is quite a useful thing to have, but it's still ultimately different from a generic state dump/load mechanism.

For one, a dump/load mechanism should handle at least DAGs (non-looping structures where individual objects are referenced multiple times) and preferably arbitrary object graphs. That's why using JX, modifications or not, won't ultimately be good in the debugger because it loses the information on object identities and doesn't support arbitrary graphs.

Debugger heap walking would preserve that identity information correctly. It'd then be up to the debug client to figure out how to pull in the graph - level by level as the user expands nodes, or as an entire graph, etc. But it'd be nice to do generic graph handling in the debug client and not on the potentially low capability target.

For non-debugger use a generic dump/load handling graphs would be useful but it would be a separate mechanism with probably quite little code sharing with JSON/JX code. It would need to represent property attributes, etc.

**Comment by fatcerberus** [comment]
Yeah, so it turns out that each BattleUnit holds a backreference to the battle engine object, which in turn holds ALL of the units in the battle, plus the Session data which is quite a large object itself.  Long story short, nearly the entire game state is reachable through a single BattleUnit object.

I'll agree that JX is far from optimal for my use case, but at least for right now, it's all I have if I want objects to display meaningful values in the debugger. :-)

**Comment by svaarala** [comment]
Right, if you want a quick fix for that, make that backreference non-enumerable using `Object.defineProperty(obj, 'backRef', { enumerable: false })`.

**Comment by fatcerberus** [comment]
Thanks for the quick hack, this seems adequate until you can come up with a better solution. :)

**Comment by svaarala** [comment]
Well there really isn't much of a better solution unless there's an actual "visited already" map (which is very different from "currently visiting"). Not to mention that is definitely not what one wants by default, because JSON behavior is serialize objects multiple times if need be.

So supporting a "visited already" list would need a separate option, and in general adding options to JX serialization. It's still a good idea I think but not a trivial change :)

**Comment by fatcerberus** [comment]
So will this be merged as-is or is it contingent on #332?

**Comment by svaarala** [comment]
It's technically an incompatible change so at least it needs some thought.

**Comment by fatcerberus** [comment]
Hm, I don't know if I would consider this specific hack incompatible as it doesn't break any existing code (it's merely lifting a restriction), it's more like adding functionality which is allowed for minor releases under semantic versioning.  But you're the boss. :)

That said, I'm going to try to come up with a better interim solution for my debugger than JX anyway, so whatever you decide to do with this is fine.

**Comment by svaarala** [comment]
JSON.stringify() guarantees an error throw from cyclic objects. Why wouldn't that guarantee matter for some code for JX too?

**Comment by svaarala** [comment]
But then, maybe that JSON guarantee is simply there to say "don't get stuck in cycles" which would be the other, ahem, implementation option.

**Comment by fatcerberus** [comment]
Ah.  I didn't realize the error throw was specifically part of the contract.  See,I tend to view exceptions as, well, exceptional, rather than part of an interface.  I get it now, carry on.

**Comment by svaarala** [comment]
Well it's part of `JSON.stringify()` contract. But the E5 specification doesn't really differentiate between "intentional" contracts and "accidental" contracts (which is quite hard to do in a specification).

For example, if the algorithm for a certain function specifies a certain argument coercion order, is that order an intentional guarantee or just there because there's no other easy way to specify algorithm than in steps?

Similarly, I'm not sure if anyone is actually relying on the cyclic object error. I guess someone could use it for something like:

```js
function isCyclicHack(obj) {
    // Close enough :)
    try { Duktape.enc('jx', obj); } catch (e) { return true; }
    return false;
}
```

Many E5 features are actually used in a surprising way. This was interesting enough that I added a test case for it, for example: http://ecmanaut.blogspot.fi/2006/07/encoding-decoding-utf8-in-javascript.html :-)

(Granted JX is not an E5 features so there's much less existing code base and thus more leeway.)

**Comment by svaarala** [comment]
Also, I'll have to look at the issue list and see if a 1.4 release makes sense before a 2.0 release. If not, then incompatible changes can start going into master and things like this can be merged as long as we agree they're good longer term changes.

I like semantic versioning but it's sometimes a bit hard to draw the line where bug fixes start and feature changes begin ;-)

Anyway, I'm off to sleep, cu :)

**Comment by fatcerberus** [comment]
I guess this could be landed in 2.0 since it's an incompatible change.

I actually have less use for this now that heap inspection is in master, but cyclic JX would still be a nice thing to have. :)  We'd just have to figure out the most useful implementation.

**Comment by fatcerberus** [comment]
@svaarala Now might be a good time to do something with this, otherwise we won't be able to do anything with it till 3.x ;)
